### PR TITLE
susbsys: settings: fcb back-end initialization and fix coverity issues

### PR DESCRIPTION
**Pull request by nvlsianpu**
API settings_subsys_init call was changed so that it returns
error (so returns int instead of void).
Prototype of storage helper function export_func for
settings_handler::h_export was changed so that it returns error
(so returns int instead of void).
Fixed few other error handling issues by ignoring return
values.

Tests were aligned to above patches.

In f. settings_init_fcb the storage area was implemented badly:
- storage was always erased
- the only source of settings fcb instance was pointed twice.
This patch fix both issues.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

### CODE REVIEW

File: subsys/settings/src/settings_store.c [hunk 1]
```diff
@@ -45,7 +45,7 @@ void settings_dst_register(struct settings_store *cs)
 
 static void settings_load_cb(char *name, char *val, void *cb_arg)
 {
-	settings_set_value(name, val);
+	ARG_UNUSED(settings_set_value(name, val));
```
Comment (lpereira) [inline subsys/settings/src/settings_store.c]: I'd rather see ARG_UNUSED only used for arguments, and not arbitrary expressions.

File: subsys/settings/src/settings_store.c [hunk 2]
```diff
@@ -118,9 +118,9 @@ int settings_save_one(const char *name, char *value)
  * Walk through all registered subsystems, and ask them to export their
  * config variables. Persist these settings.
  */
-static void settings_store_one(char *name, char *value)
+static int settings_store_one(char *name, char *value)
 {
-	settings_save_one(name, value);
+	return settings_save_one(name, value);
```
Comment (lpereira) [inline subsys/settings/src/settings_store.c]: Now that the prototypes match, maybe get rid of `settings_store_one()` and pass `settings_save_one` to whatever uses this as a callback?

**Review by lpereira — CHANGES_REQUESTED** [review]

File: tests/subsys/settings/fcb/src/settings_test_fcb.c [hunk 3]
```diff
@@ -122,10 +122,10 @@ int c1_handle_export(void (*cb)(char *name, char *value),
 	}
 
 	settings_str_from_value(SETTINGS_INT8, &val8, value, sizeof(value));
-	cb("myfoo/mybar", value);
+	ARG_UNUSED(cb("myfoo/mybar", value));
```
Comment (jhedberg) [inline tests/subsys/settings/fcb/src/settings_test_fcb.c]: This look like slight misuse of the macro. To my understanding it's intended to say "this argument is intentionally not used in this function", whereas in your case you want to say that the return value of the cb function is intentionally ignored (if I understand this right).  A more common way of handling this (at least what I've seen elsewhere) is a simple void cast in front of the function call, accompanied by an optional preceding code comment explaining why it's safe to ignore the return value.

File: subsys/settings/src/settings_store.c [hunk 4]
```diff
@@ -91,7 +93,7 @@ static void settings_dup_check_cb(char *name, char *val, void *cb_arg)
 /*
  * Append a single value to persisted config. Don't store duplicate value.
  */
-int settings_save_one(const char *name, char *value)
+int settings_save_one(char *name, char *value)
```
Comment (jhedberg) [inline subsys/settings/src/settings_store.c]: Does this change mean that it is now invalid to pass C string literals to this function? (such as "myfoo/mybar64")

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: subsys/settings/src/settings_store.c (hunk 4 above)
Comment (nvlsianpu) [reply subsys/settings/src/settings_store.c]: Dosen't mean this - but right, would be better to keep this as 'const char * '.

**Review by jhedberg — APPROVED** [review]

**Comment by nvlsianpu** [comment]
@jhedberg - This PR should resolves #7092 as well.

**Comment by jhedberg** [comment]
Thanks a lot @nvlsianpu ! My settings_load() test works now flawlessly. One thing: you might want to add a reference to the issue I opened to the commit message of your second patch.

File: subsys/settings/src/settings_store.c [hunk 5]
```diff
@@ -45,7 +46,9 @@ void settings_dst_register(struct settings_store *cs)
 
 static void settings_load_cb(char *name, char *val, void *cb_arg)
```
Comment (lpereira) [inline subsys/settings/src/settings_store.c]: Won't this trigger a build warning because `cb_arg` is unused?

**Review by lpereira — APPROVED** [review]
I'm not much of a fan of using asserts for things that can fail in runtime.  Also, those k_panic() calls when failing to initialize the settings subsystem aren't that safe.

But these things can be fixed in the future.  LGTM, just a minor nit that, if sanitycheck is fine with it, I'm fine with it as well.

**Pull request merged**